### PR TITLE
(PUP-10985) Output facts existing only in Facter 4 via `diff` action

### PR DESCRIPTION
**Pull request by luchihoratiu**
This commit aims to improve the `puppet facts diff` CLI command by:
- showing all differences between Facter 3 and Facter 4 (including facts available only in Facter 4)
- sorting output results in alphabetical order for better human readability
- allowing users to remove certain facts from output (via the `--exclude <regex>` CLI option)
- including a CLI option to see results in a fully structured form
- adding various code improvements
---
Before:
```sh-session
➜ puppet facts diff
{
  "hypervisors.vmware.version": {
    "new_value": "ESXi 6.7",
    "old_value": ""
  }
}

➜ puppet facts diff --render-as yaml
---
hypervisors.vmware.version:
  :new_value: ESXi 6.7
  :old_value: ''

➜ puppet facts diff --render-as json
{"hypervisors.vmware.version":{"new_value":"ESXi 6.7","old_value":""}}
```
---
After:
```sh-session
➜ # # # # # # #
➜ # Example of facts available only using Facter 4 on a Ubuntu 20.04
➜ puppet facts diff
{
  "disks.sda.type": {
    "new_value": "hdd",
    "old_value": null
  },
  "disks.sr0.type": {
    "new_value": "hdd",
    "old_value": null
  },
  "disks.sr1.type": {
    "new_value": "hdd",
    "old_value": null
  },
  "hypervisors.vmware.version": {
    "new_value": "ESXi 6.7",
    "old_value": ""
  },
  "networking.interfaces.ens192.bindings6.0.scope6": {
    "new_value": "link",
    "old_value": null
  },
  "networking.interfaces.lo.bindings6.0.scope6": {
    "new_value": "host",
    "old_value": null
  }
}

➜ puppet facts diff --render-as yaml
---
disks.sda.type:
  :new_value: hdd
  :old_value:
disks.sr0.type:
  :new_value: hdd
  :old_value:
disks.sr1.type:
  :new_value: hdd
  :old_value:
hypervisors.vmware.version:
  :new_value: ESXi 6.7
  :old_value: ''
networking.interfaces.ens192.bindings6.0.scope6:
  :new_value: link
  :old_value:
networking.interfaces.lo.bindings6.0.scope6:
  :new_value: host
  :old_value:

➜ puppet facts diff --render-as json
{"disks.sda.type":{"new_value":"hdd","old_value":null},"disks.sr0.type":{"new_value":"hdd","old_value":null},"disks.sr1.type":{"new_value":"hdd","old_value":null},"hypervisors.vmware.version":{"new_value":"ESXi 6.7","old_value":""},"networking.interfaces.ens192.bindings6.0.scope6":{"new_value":"link","old_value":null},"networking.interfaces.lo.bindings6.0.scope6":{"new_value":"host","old_value":null}}


➜ # # # # # # #
➜ # --structured examples
➜ puppet facts diff --structured
{
  "disks": {
    "sda": {
      "type": {
        "new_value": "hdd",
        "old_value": null
      }
    },
    "sr0": {
      "type": {
        "new_value": "hdd",
        "old_value": null
      }
    },
    "sr1": {
      "type": {
        "new_value": "hdd",
        "old_value": null
      }
    }
  },
  "hypervisors": {
    "vmware": {
      "version": {
        "new_value": "ESXi 6.7",
        "old_value": ""
      }
    }
  },
  "networking": {
    "interfaces": {
      "ens192": {
        "bindings6": {
          "0": {
            "scope6": {
              "new_value": "link",
              "old_value": null
            }
          }
        }
      },
      "lo": {
        "bindings6": {
          "0": {
            "scope6": {
              "new_value": "host",
              "old_value": null
            }
          }
        }
      }
    }
  }
}

➜ puppet facts diff --render-as yaml --structured
---
disks:
  sda:
    type:
      :new_value: hdd
      :old_value:
  sr0:
    type:
      :new_value: hdd
      :old_value:
  sr1:
    type:
      :new_value: hdd
      :old_value:
hypervisors:
  vmware:
    version:
      :new_value: ESXi 6.7
      :old_value: ''
networking:
  interfaces:
    ens192:
      bindings6:
        0:
          scope6:
            :new_value: link
            :old_value:
    lo:
      bindings6:
        0:
          scope6:
            :new_value: host
            :old_value:

➜ puppet facts diff --render-as json --structured
{"disks":{"sda":{"type":{"new_value":"hdd","old_value":null}},"sr0":{"type":{"new_value":"hdd","old_value":null}},"sr1":{"type":{"new_value":"hdd","old_value":null}}},"hypervisors":{"vmware":{"version":{"new_value":"ESXi 6.7","old_value":""}}},"networking":{"interfaces":{"ens192":{"bindings6":{"0":{"scope6":{"new_value":"link","old_value":null}}}},"lo":{"bindings6":{"0":{"scope6":{"new_value":"host","old_value":null}}}}}}}


➜ # # # # # # #
➜ # --exclude examples
➜ puppet facts diff --exclude '^disks\..*\.type$|^hypervisors.vmware.version$'
{
  "networking.interfaces.ens192.bindings6.0.scope6": {
    "new_value": "link",
    "old_value": null
  },
  "networking.interfaces.lo.bindings6.0.scope6": {
    "new_value": "host",
    "old_value": null
  }
}

➜ puppet facts diff --structured --render-as yaml --exclude disks
---
hypervisors:
  vmware:
    version:
      :new_value: ESXi 6.7
      :old_value: ''
networking:
  interfaces:
    ens192:
      bindings6:
        0:
          scope6:
            :new_value: link
            :old_value:
    lo:
      bindings6:
        0:
          scope6:
            :new_value: host
            :old_value:

➜ puppet facts diff --structured --render-as json --exclude 'disks|networking'
{"hypervisors":{"vmware":{"version":{"new_value":"ESXi 6.7","old_value":""}}}}
```

### CODE REVIEW

File: lib/puppet/util/fact_dif.rb [hunk 1]
```diff
@@ -28,35 +37,45 @@ def search_hash(sh, path = [])
...
+      hash[path] = Hash.new unless hash.key?(path)
+      bury(*paths, value, hash[path])
+    else
+      hash[*paths] = value
     end
   end
 
   def different?(new, old)
-    if old.is_a?(String) && new.is_a?(String)
+    if old.is_a?(String) && new.is_a?(String) && (old.include?(',') || new.include?(','))
```
Comment (luchihoratiu) [inline lib/puppet/util/fact_dif.rb:67]: This seems to be here to cover cases such as below:
```sh-session
➜ facter filesystems --json
{
  "filesystems": "apfs,autofs,devfs"
}
```
The extra checks ensure that unneeded splits, array diffs and checks are avoided (we get string values in most cases but there are very few where they have `,` and the value is not already an `Array`)

File: lib/puppet/face/facts.rb [hunk 2]
```diff
@@ -2,19 +2,19 @@
...
-  memoryfree memoryfree_mb
-  memory\.swap\.available_bytes memory\.swap\.used_bytes
-  memory\.swap\.available memory\.swap\.capacity memory\.swap\.used
-  memory\.system\.available_bytes memory\.system\.used_bytes
-  memory\.system\.available memory\.system\.capacity memory\.system\.used
-  mountpoints\..*\.available.* mountpoints\..*\.capacity mountpoints\..*\.used.*
-  sp_uptime system_profiler\.uptime
-  uptime uptime_days uptime_hours uptime_seconds
-  system_uptime\.uptime system_uptime\.days system_uptime\.hours system_uptime\.seconds
+EXCLUDE_LIST = %w[ ^facterversion$
```
Comment (luchihoratiu) [inline lib/puppet/face/facts.rb:5]: This is changed to avoid breaking regex given by users via `--exclude` CLI option.

**Comment by mihaibuzgau** [comment]
jenkins please test this

**Comment by mihaibuzgau** [comment]
Ran on a PE with 31 nodes connected and using the _puppet_agent::facts_diff_ task from Puppet Agent.
The output was correctly displaed.

**Review by mihaibuzgau — APPROVED** [review]

**Pull request merged**